aten/src/ATen/core/boxing/impl/boxing.h

```diff
@@ -16,25 +16,79 @@ namespace impl {
 
 // Assume T is decayed
 template <typename T>
-using not_ok_to_box = guts::negation<guts::disjunction<
+using ok_to_box = guts::disjunction<
     std::is_constructible<IValue, T>,
     // TensorOptions are not directly constructible into IValue,
     // but torch::jit::push knows how to handle them
     std::is_same<TensorOptions, T>,
     // void returns are ok
-    std::is_same<void, T>>>;
+    std::is_same<void, T>>;
 
-// TODO boxing should be ok for all kernels. Then remove not_ok_to_box and supports_boxing.
+// TODO boxing should be ok for all kernels. Then remove ok_to_box and supports_boxing.
 
-template <class Result, class... Args>
-using supports_boxing =
+template <typename Result>
+using supports_boxing_result =
   guts::negation<guts::disjunction<
     std::is_lvalue_reference<Result>,
-    not_ok_to_box<Result>,
-    std::is_same<IntArrayRef, Result>,
-    not_ok_to_box<std::decay_t<Args>>...
+    guts::negation<ok_to_box<Result>>,
+    std::is_same<IntArrayRef, Result>
   >>;
 
+template <class Result, class... Args>
+using supports_boxing =
+  guts::conjunction<
+    supports_boxing_result<Result>,
+    ok_to_box<std::decay_t<Args>>...
+  >;
+
+template <typename T, std::enable_if_t<!c10::impl::ok_to_box<T>::value>* = nullptr>
+inline bool pushIValueOrCannotBox(std::vector<c10::IValue>& stack, const T& v) {
+  torch::jit::push(stack, "cannot box");
```

Comment:
chatted with @bhosmer earlier

Reply:
It might be better not to push a string here, IIUC this will turning into an owning string in the IValue which involves a dynamic allocation (which really isn't necessary here)

Reply:
Also, if I understand correctly, previously, if we incorrectly annotated a function as supporting full c10 dispatch, we would have gotten a compile time error saying that boxAndCallBoxedFunc did not support it. With the changes here, now we no longer get a compile time error in this case. This seems a bit dangerous, as if you use this codepath to attempt to actually call a boxed function (rather than RecordFunction), you will get garbage inputs.

Reply:
sg, keeping the function here for now, and renaming to boxArgumentsOrCannotBoxIntoStack